lib/easy.c

```diff
@@ -1346,13 +1346,31 @@ static CURLcode upkeep(struct conncache *conn_cache, void *data)
  */
 CURLcode curl_easy_upkeep(struct Curl_easy *data)
 {
+  struct conncache *conn_cache;
+
   /* Verify that we got an easy handle we can work with. */
   if(!GOOD_EASY_HANDLE(data))
     return CURLE_BAD_FUNCTION_ARGUMENT;
 
-  if(data->multi_easy) {
+  /* predict the connection cache that will next be used by the easy handle.
+     if the easy handle is currently in a multi then data->state.conn_cache
+     should point to the in-use cache. otherwise use the same logic that a
```

Comment:
> Should this logic be captured in a function or macro so that it can be changed in one place rather than having it in two places?

Here's the related code from curl_multi_add_handle:

https://github.com/curl/curl/blob/7161cb17c01dcff1dc5bf89a18437d9d729f1ecd/lib/multi.c#L579-L583

IMO this code is different enough that it's not a dupe. However there is a similar "predict" type code that I noted in #12665, I'm not sure yet if the logic can be the same and combine them into a "predict_conn_cache" function